README.md
# Flutter responsive UI

This flutter package allows you to resize font responsively and get to know what kind of device the app is running on.

## Getting Started

To use this package, add `flutter_responsive_ui` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/). 

You can use this package in multiple cases. You can get the type of the device that you're running your application on ! You can find if it's a **mobile**, a **tablet** or a **desktop** ! You can even make the deference between a **small tablet** and a **large tablet** !

You can also make your UI and font size responsive depending on the device you're running the app on. By default, you'll need to develop on a **mobile emulator** because the default scale factor are based on small screens. If you want to directly develop on **tablet or desktop emulators**, feel free to use the ```init``` function to set the scale factors you want.

## Functions
You can easily know on what device you're on using : ``` isMobile(), isTablet(), isSmallTablet(), isLargeTablet() and isDesktop() ```.

To get a responsive double, use ```getResponsiveSize()``` and to get a smaller double, use ```getSmallerResponsiveSize()```. And to get the amount of column on a GridView, you can use ```responsiveGridColumnCount()```.

## Example

```dart
import 'package:flutter/material.dart';
import 'package:flutter_responsive_ui/flutter_responsive_ui.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    /// You can init all the parameters you want here :
    ResponsiveSize.initValues(
      desktopScaleFactor: 4,
      smallTabletSmallerScaleFactor: 1.17,
    );
    return MaterialApp(
      home: MyHomePage(title: 'Responsive UI'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            /// You can use the other functions to test if you're on big or small tablet, or just tablet or on desktop
            ResponsiveSize.isMobile(context: context)
                ? Text("We're on mobile")
                : Text("We're not on mobile"),

            /// Here we use the function getResponsiveSize to change the size of the text depending on the device
            Text("This text has a responsive font size !",
                style: TextStyle(
                    fontSize: ResponsiveSize.getResponsiveSize(context: context, size: 15))),

            /// Here is the same as above, but the scale factor will be lower
            Text("This text has a smaller responsive font size !",
                style: TextStyle(
                    fontSize:
                        ResponsiveSize.getSmallerResponsiveSize(context: context, size: 15))),
            Container(
              height: ResponsiveSize.getResponsiveSize(context: context, size: 500),
              child: GridView.count(
                crossAxisSpacing: ResponsiveSize.getResponsiveSize(context: context, size: 10),
                mainAxisSpacing:
                    ResponsiveSize.getSmallerResponsiveSize(context: context, size: 10),

                /// Here we use the responsive package to get a responsive cross axis column numbers
                /// We can also specify the starter amount (2 by default on mobile) using the "startColCount" parameter
                /// to get more columns if needed.
                crossAxisCount:
                    ResponsiveSize.responsiveGridColumnCount(context: context),
                children: [
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                ],
              ),
            )
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

CHANGELOG.md
## [0.1.2] - 25 August 2020

* Small bug fixes

## [0.1.1] - 25 August 2020

* BREAKING CHANGES - Context no more required but param needs name !

## [0.0.2] - 23 August 2020

* Formatting dart files !

## [0.0.1] - 23 August 2020

* First release !


lib/flutter_responsive_ui.dart
library flutter_responsive_ui;

import 'package:flutter/material.dart';

class ResponsiveSize {
  /// Here are the default values for the mobile size limit
  /// It means the limit amount of points where we stop considering it as a mobile, small tablet etc...
  static double _mobileSizeLimit = 450;
  static double _smallTabletSizeLimit = 850;
  static double _largeTabletSizeLimit = 1200;

  /// Here are the default values for the normal scale factor
  static double _mobileScaleFactor = 1.0;
  static double _smallTabletScaleFactor = 1.4;
  static double _largeTabletScaleFactor = 1.7;
  static double _desktopScaleFactor = 2;

  /// Here are the default values for the smaller scale factor
  static double _mobileSmallerScaleFactor = 1.0;
  static double _smallTabletSmallerScaleFactor = 1.15;
  static double _largeTabletSmallerScaleFactor = 1.4;
  static double _desktopSmallerScaleFactor = 1.65;

  static void initValues({
    /// All values can be set and they don't NEED to be set
    double mobileSizeLimit = 450,
    double smallTabletSizeLimit = 850,
    double largeTabletSizeLimit = 1200,
    double mobileScaleFactor = 1.0,
    double smallTabletScaleFactor = 1.4,
    double largeTabletScaleFactor = 1.7,
    double desktopScaleFactor = 2,
    double mobileSmallerScaleFactor = 1.0,
    double smallTabletSmallerScaleFactor = 1.15,
    double largeTabletSmallerScaleFactor = 1.4,
    double desktopSmallerScaleFactor = 1.65,
  }) {
    /// The initValues function is used to set the default values when the app stars
    _mobileSizeLimit = mobileSizeLimit;
    _smallTabletSizeLimit = smallTabletSizeLimit;
    _largeTabletSizeLimit = largeTabletSizeLimit;

    _mobileScaleFactor = mobileScaleFactor;
    _smallTabletScaleFactor = smallTabletScaleFactor;
    _largeTabletScaleFactor = largeTabletScaleFactor;
    _desktopScaleFactor = desktopScaleFactor;

    _mobileSmallerScaleFactor = mobileSmallerScaleFactor;
    _smallTabletSmallerScaleFactor = smallTabletSmallerScaleFactor;
    _largeTabletSmallerScaleFactor = largeTabletSmallerScaleFactor;
    _desktopSmallerScaleFactor = desktopSmallerScaleFactor;
  }

  /// The function that return whether or not it's a mobile
  static bool isMobile({BuildContext context}) {
    return MediaQuery.of(context).size.shortestSide < _mobileSizeLimit;
  }

  /// The function that return whether or not it's a tablet (small and large)
  static bool isTablet({BuildContext context}) {
    double shortestSide = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    return (shortestSide >= _mobileSizeLimit &&
        shortestSide < _largeTabletSizeLimit);
  }

  /// The function that return whether or not it's a small tablet
  static bool isSmallTablet({BuildContext context}) {
    double shortestSide = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    return (shortestSide >= _mobileSizeLimit &&
        shortestSide < _smallTabletSizeLimit);
  }

  /// The function that return whether or not it's a large tablet
  static bool isLargeTablet({BuildContext context}) {
    double shortestSide = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    return (shortestSide >= _smallTabletSizeLimit &&
        shortestSide < _largeTabletSizeLimit);
  }

  /// The function that return whether or not it's a desktop
  static bool isDesktop({BuildContext context}) {
    return (MediaQuery.of(context).size.shortestSide >= _largeTabletSizeLimit);
  }

  /// The function that return the responsive size
  static double getResponsiveSize(
      {BuildContext context, @required double size}) {
    double shortestSideSize = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    if (shortestSideSize < _mobileSizeLimit) {
      return size * _mobileScaleFactor;
    } else if (shortestSideSize >= _mobileSizeLimit &&
        shortestSideSize < _smallTabletSizeLimit) {
      return size * _smallTabletScaleFactor;
    } else if (shortestSideSize >= _smallTabletSizeLimit &&
        shortestSideSize < _largeTabletSizeLimit) {
      return size * _largeTabletScaleFactor;
    } else {
      return size * _desktopScaleFactor;
    }
  }

  /// The function that return the smaller responsive size
  static double getSmallerResponsiveSize(
      {BuildContext context, @required double size}) {
    double shortestSideSize = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    if (shortestSideSize < _mobileSizeLimit) {
      return size * _mobileSmallerScaleFactor;
    } else if (shortestSideSize >= _mobileSizeLimit &&
        shortestSideSize < _smallTabletSizeLimit) {
      return size * _smallTabletSmallerScaleFactor;
    } else if (shortestSideSize >= _smallTabletSizeLimit &&
        shortestSideSize < _largeTabletSizeLimit) {
      return size * _largeTabletSmallerScaleFactor;
    } else {
      return size * _desktopSmallerScaleFactor;
    }
  }

  /// The function that return the responsive grid item. You can choose the starting column count.
  static int responsiveGridColumnCount(
      {BuildContext context, int startColCount = 2}) {
    double shortestSideSize = context != null
        ? MediaQuery.of(context).size.shortestSide
        : WidgetsBinding.instance.window.physicalSize.width /
            WidgetsBinding.instance.window.devicePixelRatio;
    if (shortestSideSize < _mobileSizeLimit) {
      return startColCount;
    } else if (shortestSideSize >= _mobileSizeLimit &&
        shortestSideSize < _smallTabletSizeLimit) {
      return startColCount + 1;
    } else if (shortestSideSize >= _smallTabletSizeLimit &&
        shortestSideSize < _largeTabletSizeLimit) {
      return startColCount + 2;
    } else {
      return startColCount + 3;
    }
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_responsive_ui/flutter_responsive_ui.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    /// You can init all the parameters you want here :
    ResponsiveSize.initValues(
      desktopScaleFactor: 4,
      smallTabletSmallerScaleFactor: 1.17,
    );
    return MaterialApp(
      home: MyHomePage(title: 'Responsive UI'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            /// You can use the other functions to test if you're on big or small tablet, or just tablet or on desktop
            ResponsiveSize.isMobile(context: context)
                ? Text("We're on mobile")
                : Text("We're not on mobile"),

            /// Here we use the function getResponsiveSize to change the size of the text depending on the device
            Text("This text has a responsive font size !",
                style: TextStyle(
                    fontSize: ResponsiveSize.getResponsiveSize(
                        context: context, size: 15))),

            /// Here is the same as above, but the scale factor will be lower
            Text("This text has a smaller responsive font size !",
                style: TextStyle(
                    fontSize: ResponsiveSize.getSmallerResponsiveSize(
                        context: context, size: 15))),
            Container(
              height:
                  ResponsiveSize.getResponsiveSize(context: context, size: 500),
              child: GridView.count(
                crossAxisSpacing: ResponsiveSize.getResponsiveSize(
                    context: context, size: 10),
                mainAxisSpacing: ResponsiveSize.getSmallerResponsiveSize(
                    context: context, size: 10),

                /// Here we use the responsive package to get a responsive cross axis column numbers
                /// We can also specify the starter amount (2 by default on mobile) using the "startColCount" parameter
                /// to get more columns if needed.
                crossAxisCount:
                    ResponsiveSize.responsiveGridColumnCount(context: context),
                children: [
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                  Container(
                    color: Colors.grey,
                    child: Center(
                      child: Text(
                          "This grid view has a responsive crossAxisCount !"),
                    ),
                  ),
                ],
              ),
            )
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
